### ISSUE TITLE
Updated greyhole on ubuntu 16.04. Errors on run

### ISSUE THREAD

**Issue by Burden50** [gboudreau/Greyhole#171]
```
Feb 24 12:36:45 ERROR initialize: PHP Fatal Error: Uncaught Exception: SQLSTATE[HY000]: General error: 1366 Incorrect string value: '\xCB' for column 'full_path' at row 83 in /usr/bin/greyhole:619
Stack trace:
#0 /usr/bin/greyhole(900): DB::execute('ALTER TABLE `ta...')
#1 /usr/bin/greyhole(699): DB::migrate_11_varchar()
#2 /usr/bin/greyhole(600): DB::migrate()
#3 /usr/bin/greyhole(3411): DB::connect(false)
#4 /usr/bin/greyhole(3373): CommandLineHelper->getRunner()
#5 /usr/bin/greyhole(3607): CommandLineHelper->processCommandLine()
#6 {main}
  thrown; BT: greyhole[L619]
```

**Comment by gboudreau** [comment]
Could you your this in MySQL, and paste here what you get?

    DESCRIBE TABLE tasks_completed;

**Comment by Burden50** [comment]
```
mysql> DESCRIBE tasks_completed;
+-----------------+-----------------------------------------------------+------+-----+-------------------+-------+
| Field           | Type                                                | Null | Key | Default           | Extra |
+-----------------+-----------------------------------------------------+------+-----+-------------------+-------+
| id              | bigint(20) unsigned                                 | NO   |     | NULL              |       |
| action          | varchar(10)                                         | NO   |     | NULL              |       |
| share           | tinytext                                            | NO   |     | NULL              |       |
| full_path       | tinytext                                            | YES  |     | NULL              |       |
| additional_info | tinytext                                            | YES  |     | NULL              |       |
| complete        | enum('yes','no','frozen','thawed','idle','written') | NO   |     | NULL              |       |
| event_date      | timestamp                                           | NO   |     | CURRENT_TIMESTAMP |       |
+-----------------+-----------------------------------------------------+------+-----+-------------------+-------+
7 rows in set (0.17 sec)
```

**Comment by gboudreau** [comment]
SELECT * FROM information_schema.`TABLES` T,
        information_schema.`COLLATION_CHARACTER_SET_APPLICABILITY` CCSA
        WHERE CCSA.collation_name = T.table_collation
        AND T.table_schema = "greyhole"
        AND T.table_name = "tasks_completed" \G

**Comment by Burden50** [comment]
```
mysql> SELECT * FROM information_schema.`TABLES` T,     information_schema.`COLLATION_CHARACTER_SET_APPLICABILITY` CCSA     WHERE CCSA.collation_name = T.table_collation     AND T.table_schema = "service_greyhole"     AND T.table_name = "tasks_completed"\G
*************************** 1. row ***************************
     TABLE_CATALOG: def
      TABLE_SCHEMA: service_greyhole
        TABLE_NAME: tasks_completed
        TABLE_TYPE: BASE TABLE
            ENGINE: MyISAM
           VERSION: 10
        ROW_FORMAT: Dynamic
        TABLE_ROWS: 6595970
    AVG_ROW_LENGTH: 148
       DATA_LENGTH: 1877319332
   MAX_DATA_LENGTH: 281474976710655
      INDEX_LENGTH: 1024
         DATA_FREE: 894683196
    AUTO_INCREMENT: NULL
       CREATE_TIME: 2016-10-25 16:32:59
       UPDATE_TIME: 2017-02-24 06:25:04
        CHECK_TIME: 2017-01-28 06:29:20
   TABLE_COLLATION: utf8_general_ci
          CHECKSUM: NULL
    CREATE_OPTIONS:
     TABLE_COMMENT:
    COLLATION_NAME: utf8_general_ci
CHARACTER_SET_NAME: utf8
1 row in set (0.02 sec)
```

**Comment by gboudreau** [comment]
This is the query that fails. Try to run it manually; confirm it triggers the same error:

    ALTER TABLE `tasks_completed` 
        CHANGE `share` `share` VARCHAR(255) NOT NULL,
        CHANGE `full_path` `full_path` VARCHAR(255) NULL,
        CHANGE `additional_info` `additional_info` VARCHAR(255) NULL;

It's simply trying to change the `share`, `full_path` and `additional_info` columns from `TINYTEXT` to `VARCHAR(255)`; it's not trying to change the `CHARSET` of those columns, so I'm not sure why it would complain about an invalid value..?

Maybe try this query instead, which tries to make sure the new columns are really using UTF-8.

    ALTER TABLE `tasks_completed` 
        CHANGE `share` `share` VARCHAR(255) CHARACTER SET utf8 NOT NULL,
        CHANGE `full_path` `full_path` VARCHAR(255) CHARACTER SET utf8 NULL, 
        CHANGE `additional_info` `additional_info` VARCHAR(255) CHARACTER SET utf8 NULL;

Something else you can check: are those columns really UTF-8 right now?

    SELECT character_set_name FROM information_schema.`COLUMNS` 
        WHERE table_schema = "greyhole"
        AND table_name = "tasks_completed"
        AND column_name = "full_path";

**Comment by Burden50** [comment]
```
mysql> SELECT character_set_name FROM information_schema.`COLUMNS`      WHERE table_schema = "service_greyhole"     AND table_name = "tasks_completed"     AND column_name = "full_path";
+--------------------+
| character_set_name |
+--------------------+
| utf8               |
+--------------------+
1 row in set (0.00 sec)
```

```
mysql> ALTER TABLE `tasks_completed`
    ->     CHANGE `share` `share` VARCHAR(255) CHARACTER SET utf8 NOT NULL,
    ->     CHANGE `full_path` `full_path` VARCHAR(255) CHARACTER SET utf8 NULL,
    ->     CHANGE `additional_info` `additional_info` VARCHAR(255) CHARACTER SET utf8 NULL;
ERROR 1366 (HY000): Incorrect string value: '\xCB' for column 'full_path' at row 83
```


```
mysql> ALTER TABLE `tasks_completed`
    ->     CHANGE `share` `share` VARCHAR(255) NOT NULL,
    ->     CHANGE `full_path` `full_path` VARCHAR(255) NULL,
    ->     CHANGE `additional_info` `additional_info` VARCHAR(255) NULL;
ERROR 1366 (HY000): Incorrect string value: '\xCB' for column 'full_path' at row 83
```

**Comment by Burden50** [comment]
this is row 83 it seems to have some weird characters in it...

```
mysql> SELECT * FROM tasks_completed LIMIT 83,1;
+----------+--------+--------+-------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------+----------+---------------------+
| id       | action | share  | full_path                                                                                                                                             | additional_info | complete | event_date          |
+----------+--------+--------+-------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------+----------+---------------------+
| 12410853 | write  | LEVEL1 | D001/Users/Alex/AppData/Local/Packages/WinStore_cw5n1h2txyewy/LocalState/Cache/0/0-HomePageData-httpsâˆºâˆ¯âˆ¯next-services.apps.microsoft.c          | NULL            | yes      | 2017-01-23 15:07:23 |
+----------+--------+--------+-------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------+----------+---------------------+
1 row in set (0.38 sec)
```

**Comment by gboudreau** [comment]
Could you `hexdump` that row, just to be sure how those characters are encoded?

    $ mysql -u root -p -p -e 'SELECT full_path FROM greyhole.tasks_completed LIMIT 83,1' | hexdump -C

**Comment by Burden50** [comment]
```
root@SERVER:/etc# mysql -u root -p -p -e 'SELECT full_path FROM service_greyhole.tasks_completed LIMIT 83,1' | hexdump -C
Enter password:
00000000  66 75 6c 6c 5f 70 61 74  68 0a 44 30 30 31 2f 55  |full_path.D001/U|
00000010  73 65 72 73 2f 41 6c 65  78 2f 41 70 70 44 61 74  |sers/Alex/AppDat|
00000020  61 2f 4c 6f 63 61 6c 2f  50 61 63 6b 61 67 65 73  |a/Local/Packages|
00000030  2f 57 69 6e 53 74 6f 72  65 5f 63 77 35 6e 31 68  |/WinStore_cw5n1h|
00000040  32 74 78 79 65 77 79 2f  4c 6f 63 61 6c 53 74 61  |2txyewy/LocalSta|
00000050  74 65 2f 43 61 63 68 65  2f 30 2f 30 2d 48 6f 6d  |te/Cache/0/0-Hom|
00000060  65 50 61 67 65 44 61 74  61 2d 68 74 74 70 73 c3  |ePageData-https.|
00000070  a2 cb 86 c2 ba c3 a2 cb  86 c2 af c3 a2 cb 86 c2  |................|
00000080  af 6e 65 78 74 2d 73 65  72 76 69 63 65 73 2e 61  |.next-services.a|
00000090  70 70 73 2e 6d 69 63 72  6f 73 6f 66 74 2e 63 0a  |pps.microsoft.c.|
000000a0
```

```
root@SERVER:/etc# mysql -u root -p -p -e 'SELECT full_path FROM service_greyhole.tasks_completed LIMIT 82,1' | hexdump -C
Enter password:
00000000  66 75 6c 6c 5f 70 61 74  68 0a 44 30 30 31 2f 55  |full_path.D001/U|
00000010  73 65 72 73 2f 41 6c 65  78 2f 41 70 70 44 61 74  |sers/Alex/AppDat|
00000020  61 2f 4c 6f 63 61 6c 2f  50 61 63 6b 61 67 65 73  |a/Local/Packages|
00000030  2f 57 69 6e 53 74 6f 72  65 5f 63 77 35 6e 31 68  |/WinStore_cw5n1h|
00000040  32 74 78 79 65 77 79 2f  4c 6f 63 61 6c 53 74 61  |2txyewy/LocalSta|
00000050  74 65 2f 43 61 63 68 65  2f 30 2f 30 2d 50 72 6f  |te/Cache/0/0-Pro|
00000060  64 75 63 74 54 69 6c 65  45 78 74 65 6e 64 65 64  |ductTileExtended|
00000070  42 79 50 66 6e 2d 68 74  74 70 73 c3 a2 cb 0a     |ByPfn-https....|
0000007f
```

**Comment by gboudreau** [comment]
Ah, yes... LIMIT is zero-based, that last entry is the right one.
And I think that final 0xCB is the issue; that is not valid UTF-8 I think.
How is MySQL displaying it?

    mysql> SELECT * FROM tasks_completed LIMIT 82,1;

**Comment by Burden50** [comment]
```
mysql> SELECT * FROM tasks_completed LIMIT 82,1;
+----------+--------+--------+----------------------------------------------------------------------------------------------------------------------+-----------------+----------+---------------------+
| id       | action | share  | full_path                                                                                                            | additional_info | complete | event_date          |
+----------+--------+--------+----------------------------------------------------------------------------------------------------------------------+-----------------+----------+---------------------+
| 12410854 | write  | LEVEL1 | D001/Users/Alex/AppData/Local/Packages/WinStore_cw5n1h2txyewy/LocalState/Cache/0/0-ProductTileExtendedByPfn-httpsâ▒   | NULL            | yes      | 2017-01-23 15:07:23 |
+----------+--------+--------+----------------------------------------------------------------------------------------------------------------------+-----------------+----------+---------------------+
1 row in set (0.42 sec)
```

**Comment by gboudreau** [comment]
That is definitely truncated... No idea how greyhole was able to insert that, as this is not valid UTF-8...
Maybe this was just a one time fluke... What if you delete that row, then try to restart greyhole ?

    DELETE FROM tasks_completed WHERE id = 12410854;

**Comment by Burden50** [comment]
I got it running again it was very much an exercise in brute force.

1) restart greyhole
2) check log for row with issue 
3) delete row 
4) repeat until service starts

**Comment by gboudreau** [comment]
Are you somehow able to write similar files to your shares?
I'd be curious to see if there is an issue with the new columns now that the DB migration happened.

**Comment by Burden50** [comment]
I'm currently waiting for greyhole to chew though the backlog when it catches up I can try to write those files again and see if it still has issues.

**Comment by Burden50** [comment]
The backlog cleared and I tried to copy the file again it crashed but didnt seem to take down greyhole I restarted the service and it is running an fsck I will let you know the results. 

```
Feb 25 16:26:27 INFO fsck: /mnt/ARCTIC/DISKS/HD01/.landingzone/LEVEL1/D001/Users/Alex/AppData/Local/Packages/WinStore_cw5n1h2txyewy/LocalState/Cache/0/0-HomePageData-https∺∯∯next-services.apps.microsoft.com∯browse∯6.3.9600-0∯788∯en-US_en-US.en.ja.it-IT.it.fr-FR.fr.de-DE.de.es-ES.es-CL.es∯c∯CA∯cp∯10005001∯HomePageData∯pc∯0∯pt∯x64∯hw∯1∯lf∯1∯os∯Core∯Retail-.dat is a file (not a symlink). Adding a new 'write' pending task for that file.
Feb 25 16:26:27 ERROR fsck: PHP Fatal Error: Uncaught Exception: SQLSTATE[22001]: String data, right truncated: 1406 Data too long for column 'full_path' at row 1 in /usr/bin/greyhole:619
Stack trace:
#0 /usr/bin/greyhole(623): DB::execute('INSERT INTO tas...', Array)
#1 /usr/bin/greyhole(5543): DB::insert('INSERT INTO tas...', Array)
#2 /usr/bin/greyhole(5463): gh_fsck_file('/mnt/ARCTIC/DIS...', '0-HomePageData-...', 'file', 'landing_zone', 'LEVEL1', false)
#3 /usr/bin/greyhole(5461): gh_fsck('/mnt/ARCTIC/DIS...', 'LEVEL1', false)
#4 /usr/bin/greyhole(5461): gh_fsck('/mnt/ARCTIC/DIS...', 'LEVEL1', false)
#5 /usr/bin/greyhole(5461): gh_fsck('/mnt/ARCTIC/DIS...', 'LEVEL1', false)
#6 /usr/bin/greyhole(5461): gh_fsck('/mnt/ARCTIC/DIS...', 'LEVEL1', false)
#7 /usr/bin/greyhole(5461): gh_fsck('/mnt/ARCTIC/DIS...', 'LEVEL1', false)
#8 /usr/bin/greyhole(5461): gh_fsck('/mnt/ARCTIC/DIS...', 'LEVEL1', false)
#9 /usr/bin/greyhole(5461): gh_fsck('/mnt/ARCTIC/DIS...', 'LEVEL1', false)
#10 /usr/bin/greyhole(5461): gh_fsck('/mnt/ARCTIC/DIS...', 'LEVEL1', false)
#11 /; BT: greyhole[L619]
```

**Comment by gboudreau** [comment]
That file is way over the 255-characters limit Greyhole imposes on path+filename.
Even before migrating to VARCHAR(255), TINYTEXT fields do have the same limit (255 characters). That file would have been an issue even before the migration.

I'll make a change in Greyhole to catch this error, and change the appropriate fields to larger fields, so that only users who actually hit the limit will get the larger fields.

**Comment by Burden50** [comment]
Sounds Good! It's not even a file I was planning on keeping but i'm glad I could find an edge case :) 
Thanks again and keep up the good work greyhole is a solid piece of software!

**Comment by gboudreau** [comment]
Try 0.10.3; you should get larger columns automatically, when Greyhole see its first file with a length > 255 characters.

**Comment by Burden50** [comment]
I'm running this in production is it pushed to the Debian release?

**Comment by gboudreau** [comment]
Yes.

**Comment by Burden50** [comment]
After the fsck completed greyhole errored again now I cant get it to start 

```
Feb 28 13:38:00 INFO daemon: Greyhole (version 0.10.3) daemon started.
Feb 28 13:38:00 INFO daemon: Optimizing MySQL tables...
Feb 28 13:38:00 DEBUG daemon: Loading metadata store backup directories...
Feb 28 13:38:00 DEBUG daemon:   Found 2 directories in the settings table.
Feb 28 13:38:02 DEBUG daemon: Checking symlink at /usr/lib/x86_64-linux-gnu/samba/vfs/greyhole.so...
Feb 28 13:38:02 DEBUG daemon:   Is OK.
Feb 28 13:38:04 DEBUG read_smb_spool: Processing Samba spool...
Feb 28 13:38:04 ERROR read_smb_spool: PHP Fatal Error: Uncaught Exception: SQLSTATE[22001]: String data, right truncated: 1406 Data too long for column 'full_path' at row 1 in /usr/bin/greyhole:624
Stack trace:
#0 /usr/bin/greyhole(628): DB::execute('INSERT INTO tas...', Array)
#1 /usr/bin/greyhole(5114): DB::insert('INSERT INTO tas...', Array)
#2 /usr/bin/greyhole(3626): parse_samba_spool()
#3 /usr/bin/greyhole(3601): DaemonRunner->initialize()
#4 /usr/bin/greyhole(3635): DaemonRunner->run()
#5 {main}
  thrown; BT: greyhole[L624]

```

**Comment by gboudreau** [comment]
Can you see this in your Greyhole log?

> Error during MySQL query: [...]. Will now try to use larger full_path columns.

**Comment by gboudreau** [comment]
If not, can you edit /usr/bin/greyhole, around line 619, to see why it doesn't try to fix your columns?

    if ($error[1] == 1406 && $attempt_repair) {

Add a log line above it:

    Log::debug("DBG: error code = " . $error[1] . "; attempt_repair = " . ($attempt_repair ? "yes" : "no"));
    if ($error[1] == 1406 && $attempt_repair) {

**Comment by Burden50** [comment]
Nope cant see that line anywhere in the log for the last two days. 

I added that line and one of my own "TEST" which prints the $q var. 

```
Feb 28 13:53:18 DEBUG read_smb_spool: DBG: error code = ; attempt_repair = yes
Feb 28 13:53:18 DEBUG read_smb_spool: TEST: INSERT INTO tasks SET action = :action, share = :share, full_path = :full_path, additional_info = :additional_info, complete = :complete
Feb 28 13:53:18 ERROR read_smb_spool: PHP Fatal Error: Uncaught Exception: SQLSTATE[22001]: String data, right truncated: 1406 Data too long for column 'full_path' at row 1 in /usr/bin/greyhole:627
Stack trace:
```

**Comment by gboudreau** [comment]
Grr. Empty error codes...
Let's see what part of `$error` we could use:

    Log::debug("DBG: error = " . json_encode($error) . "; attempt_repair = " . ($attempt_repair ? "yes" : "no"));

**Comment by Burden50** [comment]
`Feb 28 14:06:04 DEBUG read_smb_spool: DBG: error = ["00000",null,null]; attempt_repair = yes
`

**Comment by gboudreau** [comment]
OK, what about the `$e` PDOException then:

    Log::debug("DBG: error = " . var_export($e, TRUE) . "; attempt_repair = " . ($attempt_repair ? "yes" : "no"));

**Comment by Burden50** [comment]
```
Feb 28 14:10:36 DEBUG read_smb_spool: DBG: error = PDOException::__set_state(array(
   'message' => 'SQLSTATE[22001]: String data, right truncated: 1406 Data too long for column \'full_path\' at row 1',
   'string' => '',
   'code' => '22001',
   'file' => '/usr/bin/greyhole',
   'line' => 610,
   'trace' =>
  array (
    0 =>
    array (
      'file' => '/usr/bin/greyhole',
      'line' => 610,
      'function' => 'execute',
      'class' => 'PDOStatement',
      'type' => '->',
      'args' =>
      array (
      ),
    ),
    1 =>
    array (
      'file' => '/usr/bin/greyhole',
      'line' => 634,
      'function' => 'execute',
      'class' => 'DB',
      'type' => '::',
      'args' =>
      array (
        0 => 'INSERT INTO tasks SET action = :action, share = :share, full_path = :full_path, additional_info = :additional_info, complete = :complete',
        1 =>
        array (
          'action' => 'unlink',
          'share' => 'LEVEL1',
          'full_path' => 'D001/Users/Alex/AppData/Local/Packages/WinStore_cw5n1h2txyewy/LocalState/Cache/0/0-ProductTileExtendedByProductGuid-https∺∯∯next-services.apps.microsoft.com∯browse∯6.3.9600-0∯788∯en-US_en-US.en.ja.it-IT.it.fr-FR.fr.de-DE.de.es-ES.es-CL.es∯c∯CA∯cp∯10005001∯Apps∯54C27690-1F6B-40B0-B561-72DC76E67D02.dat',
          'additional_info' => NULL,
          'complete' => 'yes',
        ),
      ),
    ),
    2 =>
    array (
      'file' => '/usr/bin/greyhole',
      'line' => 5120,
      'function' => 'insert',
      'class' => 'DB',
      'type' => '::',
      'args' =>
      array (
        0 => 'INSERT INTO tasks SET action = :action, share = :share, full_path = :full_path, additional_info = :additional_info, complete = :complete',
        1 =>
        array (
          'action' => 'unlink',
          'share' => 'LEVEL1',
          'full_path' => 'D001/Users/Alex/AppData/Local/Packages/WinStore_cw5n1h2txyewy/LocalState/Cache/0/0-ProductTileExtendedByProductGuid-https∺∯∯next-services.apps.microsoft.com∯browse∯6.3.9600-0∯788∯en-US_en-US.en.ja.it-IT.it.fr-FR.fr.de-DE.de.es-ES.es-CL.es∯c∯CA∯cp∯10005001∯Apps∯54C27690-1F6B-40B0-B561-72DC76E67D02.dat',
          'additional_info' => NULL,
          'complete' => 'yes',
        ),
      ),
    ),
    3 =>
    array (
      'file' => '/usr/bin/greyhole',
      'line' => 3632,
      'function' => 'parse_samba_spool',
      'args' =>
      array (
      ),
    ),
    4 =>
    array (
      'file' => '/usr/bin/greyhole',
      'line' => 3607,
      'function' => 'initialize',
      'class' => 'DaemonRunner',
      'type' => '->',
      'args' =>
      array (
      ),
    ),
    5 =>
    array (
      'file' => '/usr/bin/greyhole',
      'line' => 3641,
      'function' => 'run',
      'class' => 'DaemonRunner',
      'type' => '->',
      'args' =>
      array (
      ),
    ),
  ),
   'previous' => NULL,
   'errorInfo' =>
  array (
    0 => '22001',
    1 => 1406,
    2 => 'Data too long for column \'full_path\' at row 1',
  ),
)); attempt_repair = yes
```

**Comment by gboudreau** [comment]
OK, so let's go back to this for logging:

    Log::debug("DBG: error code = " . $error[1] . "; attempt_repair = " . ($attempt_repair ? "yes" : "no"));

And below line 613, add this if statement:

    $error = self::$handle->errorInfo();
    if (empty($error[1])) {
        $error = $stmt->errorInfo();
    }

**Comment by gboudreau** [comment]
Ah, this could be simplified to:

    $error = $e->errorInfo;

As the exception errorInfo property will contain either the handle or the statement errorInfo, depending on which is not empty.

**Comment by Burden50** [comment]
No crash now but still errors

`Feb 28 15:16:02 INFO read_smb_spool: Error during MySQL query: SQLSTATE[22001]: String data, right truncated: 1406 Data too long for column 'full_path' at row 1. Will now try to use larger full_path columns.`

**Comment by gboudreau** [comment]
That INFO log should only appear once. That triggers the DB migration that modifies your table to be able to hold much larger file names. So once that migration is complete, you should never see this log again, and the file you had issue with should be handled correctly from then on.

**Comment by Burden50** [comment]
Yes I was just about to update that on second restart it works fine :)

**Comment by gboudreau** [comment]
Fix released as 0.10.4
